### PR TITLE
Add $(CMD) prefixing with SIM_CMD_PREFIX

### PR DESCRIPTION
**Pull request by martijnbastiaan**
Can be used to add environment variables or prefixed commands to
invocations of $(CMD). This feature has been motivated by issue #644 to
work around issues with conflicting dynamic libraries. In these
situations users can set `SIM_CMD_PREFIX := LD_PRELOAD="foo.so bar.so"` to
force a particular library.

### CODE REVIEW

**Review by ktbarrett — COMMENTED** [review]
I think this is generally useful, but could be even more useful. I think this variable should appear just before the `$(CMD)`. It can be used for `LD_PRELOAD`; `LD_DEBUG`; `LD_LIBRARY_PATH`, `PYTHONPATH` and `PATH`; but can even be used for prepending `gdb`, `valgrind`, `perf`, etc. expanding on and nullifying the `debug` target. I proposed this in #2016.

Thoughts?

**Comment by martijnbastiaan** [comment]
Hmm, I'm not sure. On the one hand, having `COCOTB_DEBUG_CMD` would solve my issue related to having to instert `LD_PRELOAD` in order for my Xcelium installation to properly work in combination with cocotb. On the other hand, it would feel a bit like abusing the variable as the name implies it's used for debugging.

On the position of `SIM_ENV`: I purposely moved it as far forward as possible to make sure it wouldn't override any environment variables set by cocotb (although I see I failed for the very first file listed in this PR :see_no_evil:). 

My proposal would be to implement both:

1. `SIM_ENV` as far forward as possible
2. `COCOTB_DEBUG_CMD` right before `$(CMD)`

I'm happy to take on that task. If implementing both doesn't make sense to you, let's go with (2).

**Comment by themperek** [comment]
This cannot be done by adding `export  LD_PRELOAD="foo.so bar.so"` in user Makefile?

**Comment by martijnbastiaan** [comment]
That gives me a slew of warnings of all other commands now running with `LD_PRELOAD`. When it reaches the Xcelium invocation it fails in a way as if it wasn't given the `LD_PRELOAD`. Ideally, I'd pass it to Xcelium and Xcelium only.

**Comment by ktbarrett** [comment]
> On the other hand, it would feel a bit like abusing the variable as the name implies it's used for debugging.

Feel free to pick a different name. `CMD_PREFIX`, `PRECMD`, etc. 

>  If implementing both doesn't make sense to you, let's go with (2).

I think the makefiles are too knob-happy as it stands. If this can be one feature (and I think it can), it should be.

**Comment by martijnbastiaan** [comment]
`CMD_PREFIX` sounds good to me. Will implement!

File: documentation/source/building.rst
Comment (martijnbastiaan) [inline documentation/source/building.rst:305]: I've set it to `1.5.3`, but we might want to wait on a major release?

**Comment by martijnbastiaan** [comment]
@ktbarrett I've updated the PR. It implements `CMD_PREFIX` now.

File: documentation/source/building.rst
Comment (ktbarrett) [reply documentation/source/building.rst:305]:
```suggestion
      .. versionadded:: 1.6.0
```

The master branch is for the next feature release. If this were to go into a patch release this commit would need to be cherry-picked into the `stable/1.5` branch. Since this is a new non-breaking feature and not a bugfix, it should go in the next feature release.

File: documentation/source/building.rst
Comment (ktbarrett) [inline documentation/source/building.rst:303]: Convention is 1 sentence per line and I don't think you should mention motivation in reference documentation.

```suggestion
      Can be used to add environment variables or prefixed commands to invocations of $(CMD).
      For example, ``CMD_PREFIX='LD_PRELOAD="foo.so bar.so"'`` can be used to force a particular library to load.
      Or, ``CMD_PREFIX="gdb --args"`` to run the simulation with the GDB debugger.
```

File: cocotb/share/makefiles/simulators/Makefile.ghdl
Comment (ktbarrett) [inline cocotb/share/makefiles/simulators/Makefile.ghdl:73]: I'm not sure this is necessary. These are just compilation steps, cocotb is never loaded here.

File: cocotb/share/makefiles/simulators/Makefile.cvc
Comment (ktbarrett) [inline cocotb/share/makefiles/simulators/Makefile.cvc:83]:
```suggestion
         $(CMD_PREFIX) gdb --args $(CMD) $(CVC_ARGS) +acc+2 -o $(SIM_BUILD)/sim.vvp +define+COCOTB_SIM=1 +loadvpi=$(shell cocotb-config --lib-name-path vpi cvc):vlog_startup_routines_bootstrap $(COMPILE_ARGS) $(EXTRA_ARGS) $(VERILOG_SOURCES)
```
I don't think GDB allows you to add environment variables before commands like this. It needs to be before the call to `gdb`.

File: cocotb/share/makefiles/simulators/Makefile.icarus
Comment (ktbarrett) [inline cocotb/share/makefiles/simulators/Makefile.icarus:70]: Icarus also has a debug target which should get the `$(CMD_PREFIX)`.

Comment (martijnbastiaan) [reply cocotb/share/makefiles/simulators/Makefile.icarus:70]: Yeah. And this `CMD_PREFIX` was added to the compilation phase too.

This makefile is a bit weird in the sense that `CMD_PREFIX` now _doesn't_ prefix `$(CMD)`. So perhaps `CMD_PREFIX` is still a misnomer.

**Comment by martijnbastiaan** [comment]
I've applied your suggestions @ktbarrett. `CMD_PREFIX` might still be a misnomer (see inline comment). Perhaps `SIM_PREFIX`?

**Comment by ktbarrett** [comment]
@martijnbastiaan It's probably fine as is, names are not as important if there is documentation. But if you want to change it, maybe `SIM_CMD_PREFIX`, `SIM_PREFIX` doesn't make it clear its for the `CMD`. Thoughts?

**Comment by ktbarrett** [comment]
And because of your other PR, you will need to rebase =)

**Comment by martijnbastiaan** [comment]
`SIM_CMD_PREFIX` it is. I'd have to look at my own misnomer for years to come mind you ;-). I'll push the change tomorrow.

**Review by ktbarrett — APPROVED** [review]

**Comment by ktbarrett** [comment]
I usually like to leave it so others can take a look before merging. So it will get merged in a couple days.

File: documentation/source/newsfragments/2615.feature.rst
Comment (cmarqu) [inline documentation/source/newsfragments/2615.feature.rst:1]:
```suggestion
Add the :make:var:`SIM_CMD_PREFIX` to supported Makefile variables, allowing users to pass environment variables and other command prefixes to simulators.
```

File: documentation/source/building.rst
Comment (ktbarrett) [reply documentation/source/building.rst:299]: Might suggest `Prefix for simulation command invocations.`

File: cocotb/share/makefiles/Makefile.sim
Comment (ktbarrett) [reply cocotb/share/makefiles/Makefile.sim:61]: Same here.

**Comment by ktbarrett** [comment]
Not sure how coverage could have changed, so I'll ignore that.

**Pull request merged**
